logdna/resource_category.go

```diff
@@ -0,0 +1,165 @@
+  var diags diag.Diagnostics
+  pc := m.(*providerConfig)
+  categoryId := d.Id()
+  categoryType := d.Get("type").(string)
+  category := categoryRequest{}
+
+  if diags = category.CreateRequestBody(d); diags.HasError() {
+    return diags
+  }
+
+  req := newRequestConfig(
+    pc,
+    "PUT",
+    fmt.Sprintf("/v1/config/categories/%s/%s", categoryType, categoryId),
+    category,
+  )
+
+  body, err := req.MakeRequest()
+  log.Printf("[DEBUG] %s %s, payload is: %s", req.method, req.apiURL, body)
+
+  if err != nil {
+    return diag.FromErr(err)
+  }
+
+  log.Printf("[DEBUG] %s %s SUCCESS. Remote resource updated.", req.method, req.apiURL)
+
+  return resourceCategoryRead(ctx, d, m)
+}
+
+func resourceCategoryRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+  var diags diag.Diagnostics
+
+  pc := m.(*providerConfig)
+  categoryId := d.Id()
+  categoryType := d.Get("type").(string)
+
+  req := newRequestConfig(
+    pc,
+    "GET",
+    fmt.Sprintf("/v1/config/categories/%s/%s", categoryType, categoryId),
+    nil,
+  )
+
+  body, err := req.MakeRequest()
+
+  log.Printf("[DEBUG] GET categories raw response body %s\n", body)
+  if err != nil {
+    diags = append(diags, diag.Diagnostic{
+      Severity: diag.Error,
+      Summary:  "Cannot read the remote categories resource",
+      Detail:   err.Error(),
+    })
+    return diags
+  }
+
+  category := categoryResponse{}
+  err = json.Unmarshal(body, &category)
+  if err != nil {
+    diags = append(diags, diag.Diagnostic{
+      Severity: diag.Error,
+      Summary:  "Cannot unmarshal response from the remote categories resource",
+      Detail:   err.Error(),
+    })
+    return diags
+  }
+  log.Printf("[DEBUG] The GET categories structure is as follows: %+v\n", category)
+
+  appendError(d.Set("name", category.Name), &diags)
+
+  return diags
+}
+
+func resourceCategoryDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+  pc := m.(*providerConfig)
+  categoryId := d.Id()
+  categoryType := d.Get("type").(string)
+
+  req := newRequestConfig(
+    pc,
+    "DELETE",
+    fmt.Sprintf("/v1/config/categories/%s/%s", categoryType, categoryId),
+    nil,
+  )
+
+  body, err := req.MakeRequest()
+  log.Printf("[DEBUG] %s %s presetalert %s", req.method, req.apiURL, body)
+
+  if err != nil {
+    return diag.FromErr(err)
+  }
+  d.SetId("")
+  return nil
+}
+
+func resourceCategory() *schema.Resource {
+  return &schema.Resource{
+    CreateContext: resourceCategoryCreate,
+    UpdateContext: resourceCategoryUpdate,
+    ReadContext:   resourceCategoryRead,
+    DeleteContext: resourceCategoryDelete,
+    Schema: map[string]*schema.Schema{
+      "name": {
+        Type:     schema.TypeString,
+        Required: true,
+      },
+      "type": {
+        Type:     schema.TypeString,
+        Optional: true,
+        Default:  "views",
```

Comment:
Have you considered setting up an `enum` for this?  I'm not sure what the validation is on the server side--if we have such validation, then not validating here is the choice.  BUT, if there's nothing on the server side, we need to make sure that these aren't bad values.

We do that here for views. https://github.com/logdna/terraform-provider-logdna/blob/c87a6cc7be8fbf70ee2531923580702d08dc80e7/logdna/resource_view.go#L16-L21

Reply:
@darinspivey We do validate all the request on the server side and type is a enum with a strict list of options. I wanted to add enum validation here as well but I couldn't fund how to do that, I checked terraform schema source code and didn't find any enum option for a schema.

Reply:
If it's server-side validation we'd want to specifically not do that here, IMO.  So you're good.  Might be nice to have a test that rejects a bad value from the server, though.